# **Intro to Machine Learning**
 

 sadas

ML: Algoritma yg bekerja dgn menggunakan data untuk belajar dan melakuka prediksi

Machine learning:
1. Supervised Learning (Memiliki target) -> Punya target / Y
    - Predictive, Regression, Classification
2. Unsupervised Learning (Hanya berkutat pada X/fitur) -> Tidak Punya target / Y
    - Clustering 
    - PCA (Principal Component Analysis) -> mereduksi dimensi, mis. 20 variable hingga 3 variable
    - Recommendation System -> Rekomendasi dengan personalized masing2 user
    - M.B.A (Market Based Analysis) -> Kombinasi antar item (cth: brp peluang beli susu & roti)

![Alt text](image.png)

## 1. ML Supervised Learning

- Numerical -> Regression
- Categorical -> Classification

Komponen dalam ML SL

- Target / Y / Dependent Variable, umumnya 1 
- Feature / X / Independent Variable, >= 1

## 1.1 Regression

Metode modelling 

R C
- Regresi Linear -- Regresi Logistik
- KNN Regressor -- KNN Classifier
- DT Regressor -- DT Classifier

## 1.2 Types of Regression

1. Simple, x=1 (punya sifat cukup mudah diinterpretasikan)
2. Multiple Regression x>1

In [5]:
#Importing library
import statsmodels.api as sm
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy  as np

In [8]:
#Load dataset

tips = sns.load_dataset('tips')
tips

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

### Use Case 1 : Simple Regression
**Memprediksi jumlah tips dari total bill**

In [11]:
#Make regression model df

tips_pred = tips[['total_bill', 'tip']]
tips_pred

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

In [13]:
# Simple 1 feature Regression model
# total_bill as X
# tip as Y

#Create regression model sm.OLS(Y, X) , Method: Ordinary Least Squared (Memprediksi error paling kecil dgn pendekatan linear) 
reg_model = sm.OLS(tips_pred['tip'], sm.add_constant(tips_pred['total_bill']))
#
hasil = reg_model.fit()

print(hasil.summary())

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.457
Model:                            OLS   Adj. R-squared:                  0.454
Method:                 Least Squares   F-statistic:                     203.4
Date:                Mon, 18 Sep 2023   Prob (F-statistic):           6.69e-34
Time:                        10:14:56   Log-Likelihood:                -350.54
No. Observations:                 244   AIC:                             705.1
Df Residuals:                     242   BIC:                             712.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9203      0.160      5.761      0.0


**Interpreting results from Key Parameters :**

- R Squared: Berapa besar pengaruh X terhadap Y (X mempengaruhi tips sebesar 45.7%)
- Prob (F Statistic): uji keseluruhan/overall apakah Beta memberikan pengaruh (p value = 6.69e-34, < 0.05, maka tolak H0 )
    - H0: Beta = 0, Beta tidak signifikan terhadap model 
    - H1: Beta != 0, Beta signifikan terhadap model
- Uji parsial, Check P > |t| , jika < 0.05 bisa dipakai untuk persamaan prediksi
    - const coef = 0.9203
    - toal_bill coef = 0.1050


**Interpretasi:**

Y = B0 + B1X1

Y = 0.9203 + 0.1050


In [15]:
#Prediction 

def model(x):
    Y = 0.9203 + (0.1050*x)
    return Y

#Try if tips (x) = 50
print(model(50))

6.1703


In [ ]:
# Simple multiple Regression model
# total_bill as X
# tip as Y

#Create regression model sm.OLS(Y, X) , Method: Ordinary Least Squared (Memprediksi error paling kecil dgn pendekatan linear) 
reg_model = sm.OLS(tips['tip'], sm.add_constant(tips['total_bill', 'size ']))
#
hasil = reg_model.fit()

print(hasil.summary())